libraries/chain/db_block.cpp

```diff
@@ -131,8 +131,19 @@ bool database::_push_block(const signed_block& new_block)
    uint32_t skip = get_node_properties().skip_flags;
    if( !(skip&skip_fork_db) )
    {
-      /// TODO: if the block is greater than the head block and before the next maitenance interval
-      // verify that the block signer is in the current set of active witnesses.
+
+      // If the block is greater than the head block and before the next maintenance interval
+      if (new_block.block_num() > head_block_num()
```

Comment:
We need to check every block after last irreversible block, but not only those after head block.

However, for blocks those are before last maintenance interval, we aren't able to verify their witness, since we don't store the active witness list of that time in current state (not hard to change related code to store that info anyway). 

That said, active witness list in current state can be different than in the fork that contains the new block, if the fork happened before last maintenance interval. Which means it's <del>incorrect</del> less efficient (for switching to correct fork) if we drop the new block in this case.

Reply:
Let me see if I am following your thought process correctly. I see 2 points here:
1. We should change the ``if`` to look for blocks that came after the last maintenance interval instead of blocks ``> head_block_num()``
1. We should not be comparing this block's witness with the list of active witnesses, but instead compare this block's witnesses with the list of active witnesses on the fork where the new block will be placed. But we may not have enough information to do so, as we do not have the active witnesses on the fork, if the fork happened before the last maintenance interval.

Did I catch the meaning? Or am I off in the weeds?

Reply:
`1` should be last irreversible block, not last maintenance interval.

Need more discussion though. Perhaps @pmconrad will have some ideas.

Reply:
Also note that in extreme situations we can have long-lived forks that span several maintenance intervals.

Reply:
Actually this discussion can lead to questioning about the correctness of "last irreversible block" mechanism. When forks are long enough, in extreme situations, each fork will have its own LIB.

Reply:
https://github.com/EOSIO/eos/issues/2718 is related.